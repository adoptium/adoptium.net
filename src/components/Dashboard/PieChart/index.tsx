"use client"

import React from 'react';
import Highcharts from 'highcharts';
import HighchartsReact from 'highcharts-react-official';

interface PieChartProps {
    data?: Array<{ name: string; y: number }>;
    name?: string;
    showInLegend?: boolean;
    dataLabels?: boolean;
    colors?: string[] | undefined;
}

const PieChart: React.FC<PieChartProps> = ({ data, name, showInLegend = false, dataLabels = false, colors = undefined }) => {
    if (!data) return null;

    const options = {
        chart: {
            type: 'pie',
            backgroundColor: 'transparent',
            style: { fontFamily: 'inherit' },
            animation: process.env.NODE_ENV !== "test"
        },
        title: {
            text: name,
            style: { color: '#fff', fontWeight: 'bold', fontSize: '1.25rem' }
        },
        subtitle: {
            text: 'Data is from: <a href="https://api.adoptium.net/v3/stats/downloads/total" target="_blank" rel="noopener noreferrer">api.adoptium.net/v3/stats/downloads/total</a>',
            useHTML: true,
            style: { color: '#bdbdbd' }
        },
        accessibility: { enabled: false },
        tooltip: {
            backgroundColor: '#2c225a',
            borderColor: '#6ba4e7',
            style: { color: '#fff', fontWeight: 'bold' },
            formatter: function (this: unknown) {
                const ctx = this as { y?: number; point?: { name: string; percentage: number } };
                if (!ctx || !ctx.y) return '';
                const nbDownloads = Highcharts.numberFormat(ctx.y, 0, '.', ' ');
                return ctx.point?.name + '<b> ' + ctx.point?.percentage.toFixed(2) + '% </b><br/><b> ' + nbDownloads + '</b> downloads';
            }
        },
        plotOptions: {
            pie: {
                allowPointSelect: true,
                cursor: 'pointer',
                borderColor: 'rgba(255,255,255,0.1)',
                borderWidth: 2,
                dataLabels: {
                    enabled: dataLabels,
                    format: '<b>{point.name}</b>: {point.percentage:.2f} %',
                    style: { color: '#fff', fontWeight: 'bold', textOutline: 'none', fontSize: '1rem' }
                },
                colors,
                shadow: false,
                showInLegend,
            }
        },
        legend: {
            itemStyle: { color: '#fff', fontWeight: 'bold' },
            itemHoverStyle: { color: '#6ba4e7' },
        },
        series: [{
            name,
            colorByPoint: true,
            data
        }]
    };

    return (
        <div className="chart">
            <HighchartsReact highcharts={Highcharts} options={options} />
        </div>
    );
};

export default PieChart;
