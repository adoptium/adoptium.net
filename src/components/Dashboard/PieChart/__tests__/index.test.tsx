import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { describe, it, expect, afterEach } from 'vitest';
import PieChart from '../index';

afterEach(() => {
    cleanup();
});

describe('PieChart', () => {
    const mockData = [
        { name: 'Github', y: 1234 },
        { name: 'Dockerhub', y: 5678 },
        { name: 'Other', y: 100 },
    ];

    it('renders nothing if no data', () => {
        const { container } = render(<PieChart name="Test Pie" />);
        expect(container.firstChild).toBeNull();
    });

    it('renders the chart container and passes correct props', () => {
        render(<PieChart data={mockData} name="Test Pie" />);
        // The chart should render an SVG or at least a div with class 'highcharts-container' or similar
        expect(screen.getByText('Test Pie')).toBeInTheDocument();
    });

    it('renders with legend and dataLabels', () => {
        render(<PieChart data={mockData} name="Test Pie" showInLegend dataLabels />);
        expect(screen.getByText('Test Pie')).toBeInTheDocument();
    });

    it('renders with custom colors', () => {
        render(<PieChart data={mockData} name="Test Pie" colors={["#fff", "#000"]} />);
        expect(screen.getByText('Test Pie')).toBeInTheDocument();
    });
});
