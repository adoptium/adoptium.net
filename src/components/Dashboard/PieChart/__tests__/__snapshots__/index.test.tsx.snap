// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PieChart > matches snapshot 1`] = `
<div
  class="chart"
>
  <div
    data-highcharts-chart="3"
    style="overflow: hidden;"
  >
    <div
      class="highcharts-container "
      dir="ltr"
      style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none; padding: 0px; font-family: inherit;"
    >
      <svg
        aria-label="Test Pie"
        class="highcharts-root"
        height="400"
        role="img"
        style="font-family: inherit; font-size: 1rem;"
        version="1.1"
        viewBox="0 0 600 400"
        width="600"
        xmlns="http://www.w3.org/2000/svg"
      >
        <desc>
          Created with Highcharts 12.3.0
        </desc>
        <defs>
          <filter>
            <fedropshadow
              dx="1"
              dy="1"
              flood-color="#000000"
              flood-opacity="0.75"
              stdDeviation="2.5"
            />
          </filter>
        </defs>
        <rect
          class="highcharts-background"
          fill="transparent"
          filter="none"
          height="400"
          rx="0"
          ry="0"
          width="600"
          x="0"
          y="0"
        />
        <rect
          class="highcharts-plot-background"
          fill="none"
          filter="none"
          height="375"
          width="580"
          x="10"
          y="10"
        />
        <rect
          class="highcharts-plot-border"
          data-z-index="1"
          fill="none"
          height="375"
          stroke="#cccccc"
          stroke-width="0"
          width="580"
          x="10"
          y="10"
        />
        <g
          class="highcharts-series-group"
          data-z-index="3"
          filter="none"
        >
          <g
            class="highcharts-series highcharts-series-0 highcharts-pie-series highcharts-tracker"
            data-z-index="0.1"
            filter="none"
            opacity="1"
            style="cursor: pointer;"
            transform="translate(10,10) scale(1 1)"
          >
            <path
              class="highcharts-point highcharts-color-0"
              d="M 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 L 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 0 290 187.5 A 0 0 0 0 1 290 187.5 Z"
              fill="#2caffe"
              opacity="1"
              stroke="rgba(255,255,255,0.1)"
              stroke-linejoin="round"
              stroke-width="2"
              transform="translate(0,0)"
            />
            <path
              class="highcharts-point highcharts-color-1"
              d="M 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 L 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 0 290 187.5 A 0 0 0 0 1 290 187.5 Z"
              fill="#544fc5"
              opacity="1"
              stroke="rgba(255,255,255,0.1)"
              stroke-linejoin="round"
              stroke-width="2"
              transform="translate(0,0)"
            />
            <path
              class="highcharts-point highcharts-color-2"
              d="M 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 1 290 187.5 L 290 187.5 A 0 0 0 0 1 290 187.5 A 0 0 0 0 0 290 187.5 A 0 0 0 0 1 290 187.5 Z"
              fill="#00e272"
              opacity="1"
              stroke="rgba(255,255,255,0.1)"
              stroke-linejoin="round"
              stroke-width="2"
              transform="translate(0,0)"
            />
          </g>
          <g
            class="highcharts-markers highcharts-series-0 highcharts-pie-series"
            data-z-index="0.1"
            opacity="1"
            transform="translate(10,10) scale(1 1)"
          />
        </g>
        <text
          class="highcharts-title"
          data-z-index="4"
          style="font-size: 1.25rem; font-weight: bold; fill: #fff;"
          text-align="center"
          text-anchor="middle"
          transform="translate(0,0) scale(1 1)"
          transform-origin="10 4"
          x="300"
          y="10"
        >
          Test Pie
        </text>
        <text
          class="highcharts-caption"
          data-z-index="4"
          style="font-size: 0.8em; fill: #666666;"
          text-align="left"
          text-anchor="start"
          transform="translate(0,0) scale(1 1)"
          transform-origin="10 3"
          x="10"
          y="387"
        />
        <g
          class="highcharts-datalabels-group"
          data-z-index="6"
        />
        <g
          class="highcharts-legend highcharts-no-tooltip"
          data-z-index="7"
          visibility="hidden"
        >
          <rect
            class="highcharts-legend-box"
            fill="none"
            filter="none"
            height="8"
            rx="0"
            ry="0"
            stroke="#999999"
            stroke-width="0"
            width="8"
            x="0"
            y="0"
          />
          <g
            data-z-index="1"
          >
            <g />
          </g>
        </g>
        <text
          class="highcharts-credits"
          data-z-index="8"
          style="cursor: pointer; font-size: 0.6em; fill: #999999;"
          text-align="right"
          text-anchor="end"
          x="590"
          y="395"
        >
          Highcharts.com
        </text>
      </svg>
      <span
        class="highcharts-subtitle"
        data-z-index="4"
        style="position: absolute; font-family: inherit; font-size: 0.8em; white-space: nowrap; text-overflow: ellipsis; overflow: hidden; margin-left: 0px; margin-top: 0px; left: 300px; top: 8px; text-align: center; transform-origin: 0px 2px; color: rgb(189, 189, 189); transform: scale(1,1);"
        text-align="center"
        transform-origin="10 3"
      >
        Data is from: 
        <a
          href="https://api.adoptium.net/v3/stats/downloads/total"
          target="_blank"
        >
          api.adoptium.net/v3/stats/downloads/total
        </a>
      </span>
    </div>
  </div>
</div>
`;
